python_bindings/bindings.cpp

```diff
@@ -595,10 +610,12 @@ class Index {
             data_numpy_l = new hnswlib::labeltype[rows * k];
             data_numpy_d = new dist_t[rows * k];
 
+            CustomFilterFunctor idFilter((filter != nullptr)?filter:[](unsigned int id){return true;});
```

Comment:
Can we do here something so that when `filter == nullptr`
the check
https://github.com/nmslib/hnswlib/blob/687ca85e7a8c77419a8335bca55e1b00c2576788/hnswlib/hnswalg.h#L280

was true for `labelFilter` ?
It is needed to not call filter function if it is disabled (allows all ids).

Reply:
Not sure I understand. Sure it would be better to save the cpu cycle for the check if no filter function is given, but since the python binding is compiled before it is known if the search is filtered or not, I do not see a way to eliminate the run-time check.

Reply:
Thanks for the fixes!
I think of some options to adress it:
1) move the filtering template to search method as it is the only method that uses them and switch between them in python bindings code. An added advantage us that there would be possible to have several different filters via C++ templates without virtual functions.
2) Add a flag that lets the index know that there is no actual filtering so that index can skip it.

Reply:
@yurymalkov would these two approaches require changes to the files hnswlib/hnswlib.h and hnhswlib/hnswalg.h? Sorry, my C++ knowledge is limmited and I am struggeling to code a solution here although algorithmically it is clear what we want .